Usermanagement for the REST-API (Vert.x / Kotlin / Keycloak)

We have to provide permission checks not only based on the roles "create", "view", "modify" and "delete" but we should make this database dependent (databasename-create, databasename-view...)?

@JohannesLichtenberger Hey I'm getting this exception when I try to run the project any idea why ?
```
/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/bin/java "-javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=50281:/Applications/IntelliJ IDEA.app/Contents/bin" -Dfile.encoding=UTF-8 -classpath /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/deploy.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/cldrdata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/dnsns.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/jaccess.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/jfxrt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/localedata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/nashorn.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/sunec.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/sunjce_provider.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/sunpkcs11.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/ext/zipfs.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/javaws.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/jfxswt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/management-agent.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/plugin.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/lib/ant-javafx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/lib/dt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/lib/javafx-mx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/lib/jconsole.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/lib/packager.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/lib/sa-jdi.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/lib/tools.jar:/Users/ilias/Workspace/sirix/bundles/sirix-rest-api/target/classes:/Users/ilias/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib/1.3.10/kotlin-stdlib-1.3.10.jar:/Users/ilias/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib-common/1.3.10/kotlin-stdlib-common-1.3.10.jar:/Users/ilias/.m2/repository/org/jetbrains/annotations/13.0/annotations-13.0.jar:/Users/ilias/.m2/repository/org/jetbrains/kotlinx/kotlinx-coroutines-core/1.0.1/kotlinx-coroutines-core-1.0.1.jar:/Users/ilias/.m2/repository/org/jetbrains/kotlinx/kotlinx-coroutines-core-common/1.0.1/kotlinx-coroutines-core-common-1.0.1.jar:/Users/ilias/.m2/repository/io/vertx/vertx-core/3.7.0/vertx-core-3.7.0.jar:/Users/ilias/.m2/repository/io/netty/netty-common/4.1.30.Final/netty-common-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-buffer/4.1.30.Final/netty-buffer-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-transport/4.1.30.Final/netty-transport-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-handler/4.1.30.Final/netty-handler-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-codec/4.1.30.Final/netty-codec-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-handler-proxy/4.1.30.Final/netty-handler-proxy-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-codec-socks/4.1.30.Final/netty-codec-socks-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-codec-http/4.1.30.Final/netty-codec-http-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-codec-http2/4.1.30.Final/netty-codec-http2-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-resolver/4.1.30.Final/netty-resolver-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-resolver-dns/4.1.30.Final/netty-resolver-dns-4.1.30.Final.jar:/Users/ilias/.m2/repository/io/netty/netty-codec-dns/4.1.30.Final/netty-codec-dns-4.1.30.Final.jar:/Users/ilias/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.9.8/jackson-core-2.9.8.jar:/Users/ilias/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.9.8/jackson-databind-2.9.8.jar:/Users/ilias/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.9.0/jackson-annotations-2.9.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-web/3.7.0/vertx-web-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-web-common/3.7.0/vertx-web-common-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-auth-common/3.7.0/vertx-auth-common-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-bridge-common/3.7.0/vertx-bridge-common-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-lang-kotlin/3.7.0/vertx-lang-kotlin-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-lang-kotlin-coroutines/3.7.0/vertx-lang-kotlin-coroutines-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-auth-oauth2/3.7.0/vertx-auth-oauth2-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-jwt/3.7.0/vertx-jwt-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-web-client/3.7.0/vertx-web-client-3.7.0.jar:/Users/ilias/.m2/repository/io/vertx/vertx-config/3.7.0/vertx-config-3.7.0.jar:/Users/ilias/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib-jdk8/1.3.10/kotlin-stdlib-jdk8-1.3.10.jar:/Users/ilias/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib-jdk7/1.3.10/kotlin-stdlib-jdk7-1.3.10.jar:/Users/ilias/.m2/repository/io/sirix/sirix-core/0.9.4-SNAPSHOT/sirix-core-0.9.4-20191018.124639-147.jar:/Users/ilias/.m2/repository/org/xerial/snappy/snappy-java/1.1.7.2/snappy-java-1.1.7.2.jar:/Users/ilias/.m2/repository/org/magicwerk/brownies-collections/0.9.13/brownies-collections-0.9.13.jar:/Users/ilias/.m2/repository/com/google/crypto/tink/tink/1.2.1/tink-1.2.1.jar:/Users/ilias/.m2/repository/com/amazonaws/aws-java-sdk-core/1.11.166/aws-java-sdk-core-1.11.166.jar:/Users/ilias/.m2/repository/software/amazon/ion/ion-java/1.0.2/ion-java-1.0.2.jar:/Users/ilias/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-cbor/2.6.7/jackson-dataformat-cbor-2.6.7.jar:/Users/ilias/.m2/repository/com/amazonaws/aws-java-sdk-kms/1.11.166/aws-java-sdk-kms-1.11.166.jar:/Users/ilias/.m2/repository/com/amazonaws/jmespath-java/1.11.166/jmespath-java-1.11.166.jar:/Users/ilias/.m2/repository/com/google/api-client/google-api-client/1.22.0/google-api-client-1.22.0.jar:/Users/ilias/.m2/repository/com/google/oauth-client/google-oauth-client/1.22.0/google-oauth-client-1.22.0.jar:/Users/ilias/.m2/repository/com/google/http-client/google-http-client/1.22.0/google-http-client-1.22.0.jar:/Users/ilias/.m2/repository/org/apache/httpcomponents/httpclient/4.0.1/httpclient-4.0.1.jar:/Users/ilias/.m2/repository/org/apache/httpcomponents/httpcore/4.0.1/httpcore-4.0.1.jar:/Users/ilias/.m2/repository/commons-logging/commons-logging/1.1.1/commons-logging-1.1.1.jar:/Users/ilias/.m2/repository/commons-codec/commons-codec/1.3/commons-codec-1.3.jar:/Users/ilias/.m2/repository/com/google/http-client/google-http-client-jackson2/1.22.0/google-http-client-jackson2-1.22.0.jar:/Users/ilias/.m2/repository/com/google/apis/google-api-services-cloudkms/v1-rev9-1.22.0/google-api-services-cloudkms-v1-rev9-1.22.0.jar:/Users/ilias/.m2/repository/com/google/auto/service/auto-service/1.0-rc4/auto-service-1.0-rc4.jar:/Users/ilias/.m2/repository/com/google/auto/auto-common/0.8/auto-common-0.8.jar:/Users/ilias/.m2/repository/com/google/protobuf/protobuf-java/3.6.0/protobuf-java-3.6.0.jar:/Users/ilias/.m2/repository/org/json/json/20170516/json-20170516.jar:/Users/ilias/.m2/repository/io/sirix/sirix-xquery/0.9.4-SNAPSHOT/sirix-xquery-0.9.4-20191018.124727-147.jar:/Users/ilias/.m2/repository/org/jsoup/jsoup/1.11.3/jsoup-1.11.3.jar:/Users/ilias/.m2/repository/com/beust/jcommander/1.69/jcommander-1.69.jar:/Users/ilias/.m2/repository/org/aspectj/aspectjrt/1.8.10/aspectjrt-1.8.10.jar:/Users/ilias/.m2/repository/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.jar:/Users/ilias/.m2/repository/org/perfidix/perfidix/3.6.9/perfidix-3.6.9.jar:/Users/ilias/.m2/repository/commons-primitives/commons-primitives/20041207.202534/commons-primitives-20041207.202534.jar:/Users/ilias/.m2/repository/commons-math/commons-math/1.2/commons-math-1.2.jar:/Users/ilias/.m2/repository/joda-time/joda-time/2.1/joda-time-2.1.jar:/Users/ilias/.m2/repository/xmlunit/xmlunit/1.6/xmlunit-1.6.jar:/Users/ilias/.m2/repository/ch/qos/logback/logback-classic/1.2.3/logback-classic-1.2.3.jar:/Users/ilias/.m2/repository/ch/qos/logback/logback-core/1.2.3/logback-core-1.2.3.jar:/Users/ilias/.m2/repository/com/google/inject/guice/4.2.2/guice-4.2.2.jar:/Users/ilias/.m2/repository/javax/inject/javax.inject/1/javax.inject-1.jar:/Users/ilias/.m2/repository/aopalliance/aopalliance/1.0/aopalliance-1.0.jar:/Users/ilias/.m2/repository/com/google/code/gson/gson/2.8.1/gson-2.8.1.jar:/Users/ilias/.m2/repository/com/google/guava/guava/27.0-jre/guava-27.0-jre.jar:/Users/ilias/.m2/repository/com/google/guava/failureaccess/1.0/failureaccess-1.0.jar:/Users/ilias/.m2/repository/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar:/Users/ilias/.m2/repository/org/checkerframework/checker-qual/2.5.2/checker-qual-2.5.2.jar:/Users/ilias/.m2/repository/com/google/errorprone/error_prone_annotations/2.2.0/error_prone_annotations-2.2.0.jar:/Users/ilias/.m2/repository/com/google/j2objc/j2objc-annotations/1.1/j2objc-annotations-1.1.jar:/Users/ilias/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.17/animal-sniffer-annotations-1.17.jar:/Users/ilias/.m2/repository/com/google/guava/guava-testlib/27.0-jre/guava-testlib-27.0-jre.jar:/Users/ilias/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar:/Users/ilias/.m2/repository/com/github/sirixdb/brackit/brackit/0.1.5/brackit-0.1.5.jar:/Users/ilias/.m2/repository/com/github/ben-manes/caffeine/caffeine/2.7.0/caffeine-2.7.0.jar org.sirix.rest.MainKt
23:28:57,073 |-INFO in ch.qos.logback.classic.LoggerContext[default] - Found resource [logback-test.xml] at [jar:file:/Users/ilias/.m2/repository/io/sirix/sirix-core/0.9.4-SNAPSHOT/sirix-core-0.9.4-20191018.124639-147.jar!/logback-test.xml]
23:28:57,074 |-WARN in ch.qos.logback.classic.LoggerContext[default] - Resource [logback-test.xml] occurs multiple times on the classpath.
23:28:57,074 |-WARN in ch.qos.logback.classic.LoggerContext[default] - Resource [logback-test.xml] occurs at [jar:file:/Users/ilias/.m2/repository/io/sirix/sirix-xquery/0.9.4-SNAPSHOT/sirix-xquery-0.9.4-20191018.124727-147.jar!/logback-test.xml]
23:28:57,074 |-WARN in ch.qos.logback.classic.LoggerContext[default] - Resource [logback-test.xml] occurs at [jar:file:/Users/ilias/.m2/repository/io/sirix/sirix-core/0.9.4-SNAPSHOT/sirix-core-0.9.4-20191018.124639-147.jar!/logback-test.xml]
23:28:57,084 |-INFO in ch.qos.logback.core.joran.spi.ConfigurationWatchList@6dbb137d - URL [jar:file:/Users/ilias/.m2/repository/io/sirix/sirix-core/0.9.4-SNAPSHOT/sirix-core-0.9.4-20191018.124639-147.jar!/logback-test.xml] is not of type file
23:28:57,122 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - debug attribute not set
23:28:57,123 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - About to instantiate appender of type [ch.qos.logback.core.ConsoleAppender]
23:28:57,127 |-INFO in ch.qos.logback.core.joran.action.AppenderAction - Naming appender as [STDOUT]
23:28:57,131 |-INFO in ch.qos.logback.core.joran.action.NestedComplexPropertyIA - Assuming default type [ch.qos.logback.classic.encoder.PatternLayoutEncoder] for [encoder] property
23:28:57,151 |-INFO in ch.qos.logback.classic.joran.action.RootLoggerAction - Setting level of ROOT logger to ERROR
23:28:57,151 |-INFO in ch.qos.logback.core.joran.action.AppenderRefAction - Attaching appender named [STDOUT] to Logger[ROOT]
23:28:57,152 |-INFO in ch.qos.logback.classic.joran.action.ConfigurationAction - End of configuration.
23:28:57,152 |-INFO in ch.qos.logback.classic.joran.JoranConfigurator@3c9d0b9d - Registering current configuration as safe fallback point

io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8080
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:340)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused
	... 11 more
Oct 18, 2019 11:29:02 PM io.vertx.core.impl.launcher.commands.VertxIsolatedDeployer
SEVERE: Failed in deploying verticle
io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8080
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:340)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused
	... 11 more

io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8080
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:340)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused
	... 11 more
```

--- a/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/Auth.kt
+++ b/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/Auth.kt
@@ -13,7 +13,7 @@ import io.vertx.kotlin.ext.auth.isAuthorizedAwait
 /**
  * Authentication.
  */
-class Auth(private val keycloak: OAuth2Auth, private val role: String) {
+class Auth(private val keycloak: OAuth2Auth, private val role: AuthRole) {
     suspend fun handle(ctx: RoutingContext): Route {
         val token = ctx.request().getHeader(HttpHeaders.AUTHORIZATION.toString())
 
@@ -23,10 +23,10 @@ class Auth(private val keycloak: OAuth2Auth, private val role: String) {
                 "token_type" to "Bearer"
             )
         }
-
+        val database = ctx.pathParam("database")
         val user = keycloak.authenticateAwait(tokenToAuthenticate)
 
-        val isAuthorized = user.isAuthorizedAwait(role)
+        val isAuthorized = user.isAuthorizedAwait(role.databaseRole(database))
 
         if (!isAuthorized) {
             ctx.fail(HttpResponseStatus.UNAUTHORIZED.code())

--- a/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/AuthRole.kt
+++ b/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/AuthRole.kt
@@ -1,0 +1,11 @@
+package org.sirix.rest
+
+enum class AuthRole(role: String) {
+    CREATE("create"),
+    MODIFY("modify"),
+    VIEW("view"),
+    DELETE("delete");
+
+    fun keycloakRole() = "realm:$this"
+    fun databaseRole(database: String) = "realm:$database-$this"
+}

--- a/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/SirixVerticle.kt
+++ b/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/SirixVerticle.kt
@@ -91,33 +91,33 @@ class SirixVerticle : CoroutineVerticle() {
 
         // Create.
         put("/:database").consumes("application/xml").coroutineHandler {
-            Auth(keycloak, "realm:create").handle(it)
+            Auth(keycloak, AuthRole.CREATE).handle(it)
             it.next()
         }.handler(BodyHandler.create()).coroutineHandler {
             XmlCreate(location, false).handle(it)
         }
         put("/:database").consumes("application/json").coroutineHandler {
-            Auth(keycloak, "realm:create").handle(it)
+            Auth(keycloak, AuthRole.CREATE).handle(it)
             it.next()
         }.handler(BodyHandler.create()).coroutineHandler {
             JsonCreate(location, false).handle(it)
         }
 
         put("/:database/:resource").consumes("application/xml").coroutineHandler {
-            Auth(keycloak, "realm:create").handle(it)
+            Auth(keycloak, AuthRole.CREATE).handle(it)
             it.next()
         }.handler(BodyHandler.create()).coroutineHandler {
             XmlCreate(location, false).handle(it)
         }
         put("/:database/:resource").consumes("application/json").coroutineHandler {
-            Auth(keycloak, "realm:create").handle(it)
+            Auth(keycloak, AuthRole.CREATE).handle(it)
             it.next()
         }.handler(BodyHandler.create()).coroutineHandler {
             JsonCreate(location, false).handle(it)
         }
 
         post("/:database").consumes("multipart/form-data").coroutineHandler {
-            Auth(keycloak, "realm:create").handle(it)
+            Auth(keycloak, AuthRole.CREATE).handle(it)
             it.next()
         }.handler(BodyHandler.create()).coroutineHandler {
             CreateMultipleResources(location).handle(it)
@@ -128,7 +128,7 @@ class SirixVerticle : CoroutineVerticle() {
             .consumes("application/xml")
             .produces("application/xml")
             .coroutineHandler {
-                Auth(keycloak, "realm:modify").handle(it)
+                Auth(keycloak, AuthRole.MODIFY).handle(it)
                 it.next()
             }.handler(BodyHandler.create()).coroutineHandler {
                 XmlUpdate(location).handle(it)
@@ -137,7 +137,7 @@ class SirixVerticle : CoroutineVerticle() {
             .consumes("application/json")
             .produces("application/json")
             .coroutineHandler {
-                Auth(keycloak, "realm:modify").handle(it)
+                Auth(keycloak, AuthRole.MODIFY).handle(it)
                 it.next()
             }.handler(BodyHandler.create()).coroutineHandler {
                 JsonUpdate(location).handle(it)
@@ -145,51 +145,51 @@ class SirixVerticle : CoroutineVerticle() {
 
         // Get.
         get("/").produces("application/xml").coroutineHandler {
-            Auth(keycloak, "realm:view").handle(it)
+            Auth(keycloak, AuthRole.VIEW).handle(it)
             it.next()
         }.coroutineHandler {
             XmlGet(location).handle(it)
         }
         get("/").produces("application/json").coroutineHandler {
-            Auth(keycloak, "realm:view").handle(it)
+            Auth(keycloak, AuthRole.VIEW).handle(it)
             it.next()
         }.coroutineHandler {
             JsonGet(location).handle(it)
         }
 
         head("/:database/:resource").produces("application/xml").coroutineHandler {
-            Auth(keycloak, "realm:view").handle(it)
+            Auth(keycloak, AuthRole.VIEW).handle(it)
             it.next()
         }.coroutineHandler {
             XmlHead(location).handle(it)
         }
         get("/:database/:resource").produces("application/xml").coroutineHandler {
-            Auth(keycloak, "realm:view").handle(it)
+            Auth(keycloak, AuthRole.VIEW).handle(it)
             it.next()
         }.coroutineHandler {
             XmlGet(location).handle(it)
         }
         head("/:database/:resource").produces("application/json").coroutineHandler {
-            Auth(keycloak, "realm:view").handle(it)
+            Auth(keycloak, AuthRole.VIEW).handle(it)
             it.next()
         }.coroutineHandler {
             JsonHead(location).handle(it)
         }
         get("/:database/:resource").produces("application/json").coroutineHandler {
-            Auth(keycloak, "realm:view").handle(it)
+            Auth(keycloak, AuthRole.VIEW).handle(it)
             it.next()
         }.coroutineHandler {
             JsonGet(location).handle(it)
         }
 
         get("/:database").produces("application/xml").coroutineHandler {
-            Auth(keycloak, "realm:view").handle(it)
+            Auth(keycloak, AuthRole.VIEW).handle(it)
             it.next()
         }.coroutineHandler {
             XmlGet(location).handle(it)
         }
         get("/:database").produces("application/json").coroutineHandler {
-            Auth(keycloak, "realm:view").handle(it)
+            Auth(keycloak, AuthRole.VIEW).handle(it)
             it.next()
         }.coroutineHandler {
             JsonGet(location).handle(it)
@@ -199,7 +199,7 @@ class SirixVerticle : CoroutineVerticle() {
             .consumes("application/xml")
             .produces("application/xml")
             .coroutineHandler {
-                Auth(keycloak, "realm:view").handle(it)
+                Auth(keycloak, AuthRole.VIEW).handle(it)
                 it.next()
             }.handler(BodyHandler.create()).coroutineHandler {
                 XmlGet(location).handle(it)
@@ -208,7 +208,7 @@ class SirixVerticle : CoroutineVerticle() {
             .consumes("application/json")
             .produces("application/json")
             .coroutineHandler {
-                Auth(keycloak, "realm:view").handle(it)
+                Auth(keycloak, AuthRole.VIEW).handle(it)
                 it.next()
             }.handler(BodyHandler.create()).coroutineHandler {
                 JsonGet(location).handle(it)
@@ -217,7 +217,7 @@ class SirixVerticle : CoroutineVerticle() {
             .consumes("application/xml")
             .produces("application/xml")
             .coroutineHandler {
-                Auth(keycloak, "realm:view").handle(it)
+                Auth(keycloak, AuthRole.VIEW).handle(it)
                 it.next()
             }.handler(BodyHandler.create()).coroutineHandler {
                 XmlGet(location).handle(it)
@@ -226,7 +226,7 @@ class SirixVerticle : CoroutineVerticle() {
             .consumes("application/json")
             .produces("application/json")
             .coroutineHandler {
-                Auth(keycloak, "realm:view").handle(it)
+                Auth(keycloak, AuthRole.VIEW).handle(it)
                 it.next()
             }.handler(BodyHandler.create()).coroutineHandler {
                 JsonGet(location).handle(it)
@@ -234,32 +234,32 @@ class SirixVerticle : CoroutineVerticle() {
 
         // Delete.
         delete("/:database/:resource").consumes("application/xml").coroutineHandler {
-            Auth(keycloak, "realm:delete").handle(it)
+            Auth(keycloak, AuthRole.DELETE).handle(it)
             it.next()
         }.coroutineHandler {
             XmlDelete(location).handle(it)
         }
         delete("/:database/:resource").consumes("application/json").coroutineHandler {
-            Auth(keycloak, "realm:delete").handle(it)
+            Auth(keycloak, AuthRole.DELETE).handle(it)
             it.next()
         }.coroutineHandler {
             JsonDelete(location).handle(it)
         }
 
         delete("/:database").consumes("application/xml").coroutineHandler {
-            Auth(keycloak, "realm:delete").handle(it)
+            Auth(keycloak, AuthRole.DELETE).handle(it)
             it.next()
         }.coroutineHandler {
             XmlDelete(location).handle(it)
         }
         delete("/:database").consumes("application/json").coroutineHandler {
-            Auth(keycloak, "realm:delete").handle(it)
+            Auth(keycloak, AuthRole.DELETE).handle(it)
             it.next()
         }.coroutineHandler {
             JsonDelete(location).handle(it)
         }
         delete("/").coroutineHandler {
-            Auth(keycloak, "realm:delete").handle(it)
+            Auth(keycloak, AuthRole.DELETE).handle(it)
             it.next()
         }.coroutineHandler {
             Delete(location).handle(it)

--- a/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/crud/json/JsonSessionDBStore.kt
+++ b/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/crud/json/JsonSessionDBStore.kt
@@ -2,67 +2,68 @@ package org.sirix.rest.crud.json
 
 import com.google.gson.stream.JsonReader
 import io.vertx.ext.auth.User
+import org.sirix.rest.AuthRole
 import org.sirix.xquery.json.JsonDBCollection
 import org.sirix.xquery.json.JsonDBStore
 import java.nio.file.Path
 
 class JsonSessionDBStore(private val dbStore: JsonDBStore, private val user: User) : JsonDBStore by dbStore {
     override fun lookup(name: String): JsonDBCollection {
-        checkIfAuthorized(name, "view")
+        checkIfAuthorized(name, AuthRole.VIEW)
 
         return dbStore.lookup(name)
     }
 
     override fun create(name: String): JsonDBCollection {
-        checkIfAuthorized(name, "create")
+        checkIfAuthorized(name, AuthRole.CREATE)
 
         return dbStore.create(name)
     }
 
     override fun create(name: String, path: String): JsonDBCollection {
-        checkIfAuthorized(name, "create")
+        checkIfAuthorized(name, AuthRole.CREATE)
 
         return dbStore.create(name, path)
     }
 
     override fun create(name: String, path: Path): JsonDBCollection {
-        checkIfAuthorized(name, "create")
+        checkIfAuthorized(name, AuthRole.CREATE)
 
         return dbStore.create(name, path)
     }
 
     override fun create(collName: String, resourceName: String, path: Path): JsonDBCollection {
-        checkIfAuthorized(collName, "create")
+        checkIfAuthorized(collName, AuthRole.CREATE)
 
         return dbStore.create(collName, path)
     }
 
     override fun create(collName: String, resourceName: String, json: String): JsonDBCollection {
-        checkIfAuthorized(collName, "create")
+        checkIfAuthorized(collName, AuthRole.CREATE)
 
         return dbStore.create(collName, resourceName, json)
     }
 
     override fun create(collName: String, resourceName: String, json: JsonReader): JsonDBCollection {
-        checkIfAuthorized(collName, "create")
+        checkIfAuthorized(collName, AuthRole.CREATE)
 
         return dbStore.create(collName, resourceName, json)
     }
 
     override fun create(collName: String, jsonReaders: Set<JsonReader>): JsonDBCollection {
-        checkIfAuthorized(collName, "create")
+        checkIfAuthorized(collName, AuthRole.CREATE)
 
         return dbStore.create(collName, jsonReaders)
     }
 
     override fun drop(name: String) {
-        checkIfAuthorized(name, "delete")
+        checkIfAuthorized(name, AuthRole.DELETE)
 
         return dbStore.drop(name)
     }
 
-    private fun checkIfAuthorized(name: String, role: String) {
-        user.isAuthorized("realm:$role") {
+    private fun checkIfAuthorized(name: String, role: AuthRole) {
+        user.isAuthorized(role.databaseRole(name)) {
             if (!it.succeeded()) {
                 throw IllegalArgumentException("User is not allowed to $role the database $name")
             }

--- a/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/crud/xml/XmlSessionDBStore.kt
+++ b/bundles/sirix-rest-api/src/main/kotlin/org/sirix/rest/crud/xml/XmlSessionDBStore.kt
@@ -3,48 +3,49 @@ package org.sirix.rest.crud.xml
 import io.vertx.ext.auth.User
 import org.brackit.xquery.node.parser.SubtreeParser
 import org.brackit.xquery.xdm.Stream
+import org.sirix.rest.AuthRole
 import org.sirix.xquery.node.XmlDBCollection
 import org.sirix.xquery.node.XmlDBStore
 
 class XmlSessionDBStore(private val dbStore: XmlDBStore, private val user: User) : XmlDBStore by dbStore {
     override fun lookup(name: String): XmlDBCollection {
-        checkIfAuthorized(name, "view")
+        checkIfAuthorized(name, AuthRole.VIEW)
 
         return dbStore.lookup(name)
     }
 
     override fun create(name: String): XmlDBCollection {
-        checkIfAuthorized(name, "create")
+        checkIfAuthorized(name, AuthRole.CREATE)
 
         return dbStore.create(name)
     }
 
     override fun create(name: String, parser: SubtreeParser): XmlDBCollection {
-        checkIfAuthorized(name, "create")
+        checkIfAuthorized(name, AuthRole.CREATE)
 
         return dbStore.create(name, parser)
     }
 
     override fun create(name: String, parsers: Stream<SubtreeParser>): XmlDBCollection {
-        checkIfAuthorized(name, "create")
+        checkIfAuthorized(name, AuthRole.CREATE)
 
         return dbStore.create(name, parsers)
     }
 
     override fun create(dbName: String, resourceName: String, parsers: SubtreeParser): XmlDBCollection {
-        checkIfAuthorized(dbName, "create")
+        checkIfAuthorized(dbName, AuthRole.CREATE)
 
         return dbStore.create(dbName, resourceName, parsers)
     }
 
     override fun drop(name: String) {
-        checkIfAuthorized(name, "delete")
+        checkIfAuthorized(name, AuthRole.DELETE)
 
         return dbStore.drop(name)
     }
 
-    private fun checkIfAuthorized(name: String, role: String) {
-        user.isAuthorized("realm:$role") {
+    private fun checkIfAuthorized(name: String, role: AuthRole) {
+        user.isAuthorized(role.databaseRole(name)) {
             if (!it.succeeded()) {
                 throw IllegalArgumentException("User is not allowed to $role the database $name")
             }

